hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/record/TokenContextImpl.java

```diff
@@ -92,53 +89,34 @@ public <T> T writableStore(@NonNull Class<T> storeInterface) {
 
     @NonNull
     @Override
-    public <T> T userTransactionRecordBuilder(@NonNull Class<T> recordBuilderClass) {
+    public <T extends SingleTransactionRecordBuilder> T userTransactionRecordBuilder(
+            @NonNull Class<T> recordBuilderClass) {
         requireNonNull(recordBuilderClass, "recordBuilderClass must not be null");
-        return castRecordBuilder(recordListBuilder.userTransactionRecordBuilder(), recordBuilderClass);
+        return stack.getBaseBuilder(recordBuilderClass);
     }
 
     @Override
     public boolean hasChildOrPrecedingRecords() {
-        return !recordListBuilder.childRecordBuilders().isEmpty()
-                || !recordListBuilder.precedingRecordBuilders().isEmpty();
+        return stack.hasNonBaseStreamBuilder();
     }
 
     @Override
     public <T> void forEachChildRecord(@NonNull Class<T> recordBuilderClass, @NonNull Consumer<T> consumer) {
         requireNonNull(consumer, "consumer must not be null");
-        final var childRecordBuilders = recordListBuilder.childRecordBuilders();
-        final var precedingRecordBuilders = recordListBuilder.precedingRecordBuilders();
-
-        childRecordBuilders.forEach(child -> consumer.accept(castRecordBuilder(child, recordBuilderClass)));
-        precedingRecordBuilders.forEach(child -> consumer.accept(castRecordBuilder(child, recordBuilderClass)));
+        stack.forEachNonBaseBuilder(recordBuilderClass, consumer);
     }
 
     @NonNull
     @Override
-    public <T> T addPrecedingChildRecordBuilder(@NonNull Class<T> recordBuilderClass) {
-        final var result = recordListBuilder.addPreceding(configuration(), LIMITED_CHILD_RECORDS);
-        return castRecordBuilder(result, recordBuilderClass);
-    }
-
-    @NonNull
-    @Override
-    public <T> T addUncheckedPrecedingChildRecordBuilder(@NonNull Class<T> recordBuilderClass) {
-        final var result = recordListBuilder.addPreceding(configuration(), UNLIMITED_CHILD_RECORDS);
-        return castRecordBuilder(result, recordBuilderClass);
-    }
-
-    static <T> T castRecordBuilder(
-            @NonNull final SingleTransactionRecordBuilderImpl recordBuilder,
-            @NonNull final Class<T> recordBuilderClass) {
-        if (!recordBuilderClass.isInstance(recordBuilder)) {
-            throw new IllegalArgumentException("Not a valid record builder class");
-        }
-        return recordBuilderClass.cast(recordBuilder);
+    public <T extends SingleTransactionRecordBuilder> T addPrecedingChildRecordBuilder(
+            @NonNull Class<T> recordBuilderClass) {
+        final var result = stack.createIrreversiblePrecedingBuilder();
```

Comment:
When we explicitly add a preceding record without dispatching, its irreversible. Because they are used for EndOfStakingPeriod and genesis records. For other types when. we dispatch a preceding transaction, it is reversible.

Reply:
Ah, ok. A comment in the javadoc about this would probably be helpful (again, fine to do in another PR)